File: engine/argot/research/signal/phase14/filters/data_dominant.py
# engine/argot/research/signal/phase14/filters/data_dominant.py
"""Data-dominant file detector — tree-sitter structural heuristic, no I/O.

.. deprecated::
    Direct use of this module is deprecated.  Use
    ``argot.research.signal.phase14.adapters.PythonAdapter.is_data_dominant``
    instead.  This module is kept as a thin shim because external callers
    (validation scripts, experiments) may still import it directly.
"""

from __future__ import annotations

import tree_sitter_python as tspython
from tree_sitter import Language, Node
from tree_sitter import Parser as TsParser

_PY_LANGUAGE = Language(tspython.language())
_PARSER = TsParser(_PY_LANGUAGE)

# RHS node types that count as "static data literals"
_DATA_LITERAL_TYPES: frozenset[str] = frozenset({"list", "tuple", "dictionary", "set"})


def _get_rhs(node: Node) -> Node | None:
    """Return the right-hand-side child of an assignment node, or None."""
    # tree-sitter Python: assignment → left, '=', right
    # The right-hand side is the last named child when there are named children.
    # Walk children looking for a node that comes after '='
    found_eq = False
    for child in node.children:
        if found_eq:
            return child
        if child.type == "=":
            found_eq = True
    return None


def _collect_stmt_data_rows(stmts: list[Node], rows: set[int]) -> None:
    """Scan a list of sibling statement nodes and add data-literal spans to *rows*."""
    for stmt in stmts:
        target: Node | None = None

        if stmt.type == "assignment":
            target = stmt
        elif stmt.type == "expression_statement":
            for sub in stmt.children:
                if sub.type == "assignment":
                    target = sub
                    break

        if target is None:
            continue

        rhs = _get_rhs(target)
        if rhs is not None and rhs.type in _DATA_LITERAL_TYPES:
            rows.update(range(stmt.start_point[0], stmt.end_point[0] + 1))


def _extract_data_literal_lines(root: Node) -> frozenset[int]:
    """Return 0-indexed row numbers covered by data literal assignments.

    Scans both module-level statements and class-body statements so that
    locale-style provider files (e.g. ``class Provider: cities = (...)`` ) are
    correctly detected as data-dominant even when the data lives in a class body.
    """
    rows: set[int] = set()

    _collect_stmt_data_rows(root.children, rows)

    # Descend into top-level class definitions to catch class-body data tables.
    for child in root.children:
        if child.type == "class_definition":
            for sub in child.children:
                if sub.type == "block":
                    _collect_stmt_data_rows(sub.children, rows)

    return frozenset(rows)


def is_data_dominant(file_source: str, threshold: float = 0.65) -> bool:
    """Return True if the file is overwhelmingly composed of top-level data literal assignments.

    The heuristic parses the file with tree-sitter and walks module-level statements
    as well as class-body statements inside top-level class definitions.  Any
    ``assignment`` whose right-hand side is a ``list``, ``tuple``, ``dictionary``,
    or ``set`` literal contributes its line span to *data_literal_lines*.  The ratio::

        data_literal_lines / total_nonblank_lines

    is compared against *threshold* (default 0.65).  Files with a parse error or
    empty input return False.

    Args:
        file_source: Full file content as a string.
        threshold: Fraction above which the file is considered data-dominant.

    Returns:
        True if the data-literal ratio exceeds *threshold*, False otherwise.
    """
    if not file_source or not file_source.strip():
        return False

    try:
        tree = _PARSER.parse(file_source.encode("utf-8"))
    except Exception:
        return False

    if tree.root_node.has_error:
        # Tolerate minor errors — the heuristic degrades gracefully; but if the
        # tree is entirely an ERROR node, bail out.
        if tree.root_node.type == "ERROR":
            return False

    total_nonblank = sum(1 for line in file_source.splitlines() if line.strip())
    if total_nonblank == 0:
        return False

    data_rows = _extract_data_literal_lines(tree.root_node)
    ratio = len(data_rows) / total_nonblank
    return ratio > threshold

File: crates/argot-lang/src/filters/data_dominant.rs
//! Data-dominant file detector.
//!
//! A structural, I/O-free heuristic: parse with tree-sitter and walk
//! module-level statements plus top-level class bodies. Any `assignment`
//! whose right-hand side is a `list`/`tuple`/`dictionary`/`set` literal made
//! mostly of value literals contributes its line span. If that span covers
//! more than `threshold` of the non-blank lines, the file is data-dominant.

use std::collections::HashSet;
use tree_sitter::{Node, Parser, Tree};

use crate::text::splitlines;

/// RHS node types that count as "static data literals".
const DATA_LITERAL_TYPES: &[&str] = &["list", "tuple", "dictionary", "set"];

/// Named-value node types that count as literal data.
const VALUE_LITERAL_TYPES: &[&str] = &[
    "string",
    "concatenated_string",
    "integer",
    "float",
    "true",
    "false",
    "none",
    "list",
    "tuple",
    "dictionary",
    "set",
];

const VALUE_DOMINANT_THRESHOLD: f64 = 0.8;

fn parse(source: &str) -> Tree {
    let mut parser = Parser::new();
    parser
        .set_language(&tree_sitter_python::LANGUAGE.into())
        .expect("tree-sitter python grammar loads");
    parser
        .parse(source.as_bytes(), None)
        .expect("tree-sitter parse never returns None without a timeout")
}

fn children<'a>(node: Node<'a>) -> Vec<Node<'a>> {
    let mut cursor = node.walk();
    node.children(&mut cursor).collect()
}

/// True if ≥80% of the node's immediate named values are literal data.
/// Empty containers and unknown types return true (conservative — allow).
fn is_value_literal_dominant(node: Node) -> bool {
    let values: Vec<Node> = match node.kind() {
        "list" | "tuple" | "set" => children(node)
            .into_iter()
            .filter(|c| c.is_named())
            .collect(),
        "dictionary" => {
            let mut vals = Vec::new();
            for child in children(node) {
                match child.kind() {
                    "pair" => {
                        if let Some(val) = child.child_by_field_name("value") {
                            vals.push(val);
                        }
                    }
                    // splat (**x) is non-literal
                    "dictionary_splat" => vals.push(child),
                    _ => {}
                }
            }
            vals
        }
        _ => return true,
    };
    if values.is_empty() {
        return true;
    }
    let literal_count = values
        .iter()
        .filter(|v| VALUE_LITERAL_TYPES.contains(&v.kind()))
        .count();
    (literal_count as f64) / (values.len() as f64) >= VALUE_DOMINANT_THRESHOLD
}

/// Return the right-hand-side child of an assignment node — the first child
/// after the `=` token — or `None`.
fn get_rhs(node: Node) -> Option<Node> {
    let mut found_eq = false;
    for child in children(node) {
        if found_eq {
            return Some(child);
        }
        if child.kind() == "=" {
            found_eq = true;
        }
    }
    None
}

/// Scan sibling statement nodes, adding data-literal spans (0-indexed rows).
fn collect_stmt_data_rows(stmts: &[Node], rows: &mut HashSet<usize>) {
    for stmt in stmts {
        let target = if stmt.kind() == "assignment" {
            Some(*stmt)
        } else if stmt.kind() == "expression_statement" {
            children(*stmt)
                .into_iter()
                .find(|c| c.kind() == "assignment")
        } else {
            None
        };

        let Some(target) = target else { continue };

        if let Some(rhs) = get_rhs(target) {
            if DATA_LITERAL_TYPES.contains(&rhs.kind()) && is_value_literal_dominant(rhs) {
                for r in stmt.start_position().row..=stmt.end_position().row {
                    rows.insert(r);
                }
            }
        }
    }
}

/// 0-indexed row numbers covered by data-literal assignments, scanning both
/// module-level statements and top-level class bodies.
fn extract_data_literal_lines(root: Node) -> HashSet<usize> {
    let mut rows = HashSet::new();
    let top = children(root);
    collect_stmt_data_rows(&top, &mut rows);

    for child in &top {
        if child.kind() == "class_definition" {
            for sub in children(*child) {
                if sub.kind() == "block" {
                    collect_stmt_data_rows(&children(sub), &mut rows);
                }
            }
        }
    }
    rows
}

/// 1-indexed line numbers covered by data-literal assignment spans — the
/// row-granular view behind [`is_data_dominant`]. Empty on parse failure.
pub fn data_literal_lines(source: &str) -> HashSet<usize> {
    if source.is_empty() || source.trim().is_empty() {
        return HashSet::new();
    }
    let tree = parse(source);
    let root = tree.root_node();
    if root.has_error() && root.kind() == "ERROR" {
        return HashSet::new();
    }
    extract_data_literal_lines(root)
        .into_iter()
        .map(|r| r + 1)
        .collect()
}

/// True if the file is overwhelmingly top-level data-literal assignments.
pub fn is_data_dominant(source: &str, threshold: f64) -> bool {
    if source.is_empty() || source.trim().is_empty() {
        return false;
    }

    let tree = parse(source);
    let root = tree.root_node();

    // Tolerate minor errors, but bail if the whole tree is an ERROR node.
    if root.has_error() && root.kind() == "ERROR" {
        return false;
    }

    let total_nonblank = splitlines(source)
        .iter()
        .filter(|line| !line.trim().is_empty())
        .count();
    if total_nonblank == 0 {
        return false;
    }

    let data_rows = extract_data_literal_lines(root);
    let ratio = (data_rows.len() as f64) / (total_nonblank as f64);
    ratio > threshold
}

#[cfg(test)]
mod tests;
